### PR TITLE
Elg 272

### PR DESCRIPTION
**Pull request by gcleary**
Here it is. Uncompressed Rq:

```
0000000: 0A 0C 12 34 56 12 34 56 11 22 33 44 55 66 10 EA
0000010: C3 EA F5 05 1A 66 08 06 22 45 DD 9C 82 97 CD 04
0000020: B2 84 F3 90 B8 87 0A 8B C1 93 93 B8 87 0A 99 B9
0000030: CB 84 AB CD 20 F0 D3 D8 91 B8 87 0A DB 81 F8 84
0000040: A8 84 3B A1 C4 91 F5 AF 90 2D A2 C4 91 F5 AF 90
0000050: 2D 95 A0 F8 D2 AD 9B 1D AA E2 F1 90 B8 87 0A 10
0000060: CD 1C 32 0A 5C 5B 01 01 5A 5A 59 59 58 4D 18 AD
0000070: 02 0A 09 A3 0F 00 00 00 00 00 00 00 2A 16 2A 02
0000080: E5 7D 32 02 F2 74 1A 02 DB 04 22 01 01 3A 01 64
0000090: 42 02 BB 02 4A 23 0A 04 BC B7 C2 11 12 04 88 91
00000A0: BA 01 1A 01 6C 22 02 8E 11 2A 01 28 32 01 64 3A
00000B0: 02 8E 02 42 01 05 4A 01 64 00 00 00 00 00 00 00
```
The AP in the workspace:
```
Skyhook libEL DEBG: utilities.c:dump_workspace() WorkSpace: Got 10 beacons, WiFi 10, connected 5
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 0 : WiFi Age: 300        MAC 00:24:D2:E0:8E:5D rssi: -92  3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 1 : WiFi Age: 300        MAC 28:3B:82:1C:C2:32 rssi: -91  3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 2 : WiFi Age: 300        MAC 28:3B:82:64:E0:8B rssi: -1   3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 3 : WiFi Age: 300        MAC 82:6A:B0:92:DC:99 rssi: -1   3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 4 : WiFi Age: 300        MAC 28:3B:82:36:29:F0 rssi: -90  3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 5 : WiFi Age: 300        MAC EC:22:80:9E:00:DB rssi: -90  3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 6 : WiFi Age: 300        MAC B4:82:FE:A4:62:21 rssi: -89  3660 MHz vap: 1
Skyhook libEL DEBG: utilities.c:dump_vap()  Virt AP 6 : WiFi Age: 300        MAC B4:82:FE:A4:62:31 rssi: -89  3660 MHz VAP(A 3)
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 7 : WiFi Age: 300        MAC B4:82:FE:A4:62:22 rssi: -89  3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 8 : WiFi Age: 300        MAC 74:DA:DA:5E:10:15 rssi: -88  3660 MHz vap: 0
Skyhook libEL DEBG: utilities.c:dump_workspace()  Beacon 9 : WiFi Age: 300        MAC 28:3B:82:1C:71:2A rssi: -77  3660 MHz vap: 0
```
The compressed data bytes in the Rq appear to be in the neighborhood of 'A3' Index 0000073 in the buffer. A3 is interpreted as 'Substitute the 10th nibble of the MAC with 3. The mac in question should be specified by the preceding byte and should have value (6+1) ie 0x07. However, as you can see, there is no such value next to A3.

### CODE REVIEW

File: libel/beacons.c [hunk 1]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
  *
  *  @param macA pointer to the first MAC
  *  @param macB pointer to the second MAC
+ *  @param pd pointer to index of delta
```
Comment (tboinske-skyhook) [inline libel/beacons.c:48]: "pointer to _nibble_ index (0-11) of patch", right?

File: libel/beacons.c [hunk 2]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+        }
+        if ((macA[i] & 0x0F) != (macB[i] & 0x0F)) {
+            if (++num_diff > 1)
+                return 0;
+            idx_diff = i * 2 + 1;
+            smaller = macA[i] - macB[i];
+        }
     }
 
     /* MACs are similar, choose one to remove */
```
Comment (tboinske-skyhook) [inline libel/beacons.c:82]: This function in and of itself has nothing to do with "removing" anything. Perhaps this comment can just be deleted.

File: libel/beacons.c [hunk 3]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
  *  @return -1, 0 or 1
  *  return 0 when NOT similar, -1 indicates keep A, 1 keep B
+ *  if macs are similar, and index is not NULL, index is set to nibble index of difference
  */
-static int similar(uint8_t macA[], uint8_t macB[])
+static int similar(uint8_t macA[], uint8_t macB[], int *pd)
 {
     /* Return 1 (true) if OUIs are identical and no more than 1 hex digits
      * differ between the two MACs. Else return 0 (false).
      */
```
Comment (tboinske-skyhook) [inline libel/beacons.c:58]: I think the comments starting at line 50 obviate the need for this comment, which is wrong in any case.

File: libel/beacons.c [hunk 4]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
     size_t num_diff = 0; // Num hex digits which differ
+    size_t idx_diff = 0; // nibble digit which differs
     size_t i;
+    int smaller = 0;
```
Comment (tboinske-skyhook) [inline libel/beacons.c:62]: Would "sign" (or even "ret_val") be a better variable name than "smaller"?

File: libel/beacons.c [hunk 5]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+        ctx->beacon[ap].ap.mac[0], ctx->beacon[ap].ap.mac[1], ctx->beacon[ap].ap.mac[2],
+        ctx->beacon[ap].ap.mac[3], ctx->beacon[ap].ap.mac[4], ctx->beacon[ap].ap.mac[5],
+        ctx->beacon[ap].ap.rssi, ctx->beacon[ap].ap.freq);
+#endif
+
+    if (vg >= NUM_APS(ctx) || ap >= NUM_APS(ctx))
+        return false;
+
+    if ((n = nibble(ctx->beacon[ap].ap.mac, d)) == 0xff)
+        return false;
```
Comment (tboinske-skyhook) [inline libel/beacons.c:147]: Why? (Need comment.)

File: libel/beacons.c [hunk 6]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+    else
+        return ((mac[d / 2] >> 4) & 0x0F);
+}
+
+/*! \brief add an AP to Group, including any from thay APs Group
+ *
+ *  @param ctx Skyhook request context
+ *  @param vg index of virtual group mac
+ *  @param ap index of AP to add to group
+ *  @param d index of nibble that differs
```
Comment (tboinske-skyhook) [inline libel/beacons.c:110]: "... (0-11)..."

File: libel/beacons.c [hunk 7]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+#endif
+
+    if (vg >= NUM_APS(ctx) || ap >= NUM_APS(ctx))
+        return false;
+
+    if ((n = nibble(ctx->beacon[ap].ap.mac, d)) == 0xff)
+        return false;
+
+    pvg = ctx->beacon[vg].ap.vg;
+    delta = (uint8_t)d << 4 | n;
```
Comment (tboinske-skyhook) [inline libel/beacons.c:150]: Would "entry" be a better name than "delta"?

File: libel/beacons.c [hunk 8]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+    /* find end of list of APs in Virtual group. */
+    /* Leave at least one terminating 0x00 */
+    for (dup = v = 0; v < ctx->ap_vg_len - 1; v++) {
+        if ((pvg[v] & 0xF0) == 0)
+            break;
+        if (pvg[v] == delta)
+            dup = 1;
```
Comment (tboinske-skyhook) [inline libel/beacons.c:157]: Why would we expect to see a duplicate entry?

File: libel/beacons.c [hunk 9]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+    if (d & 1)
+        return (mac[d / 2] & 0x0F);
+    else
+        return ((mac[d / 2] >> 4) & 0x0F);
+}
+
+/*! \brief add an AP to Group, including any from thay APs Group
+ *
+ *  @param ctx Skyhook request context
+ *  @param vg index of virtual group mac
```
Comment (tboinske-skyhook) [inline libel/beacons.c:108]: Meaning the index of the "parent" AP (the group member with the lowest MAC address value)?

File: libel/beacons.c [hunk 10]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+ *  @param ap index of AP to add to group
```
Comment (tboinske-skyhook) [inline libel/beacons.c:109]: Meaning the "child"?

File: libel/beacons.c [hunk 11]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+        return false;
+
+    if (!dup) {
+        pvg[v] = delta;
+        ctx->beacon[vg].ap.vg_len = v + 1;
+    }
+
+    /* Add any virtual APs from the one being added (if it is a group) */
+    for (av = 0; av < ctx->ap_vg_len; av++) {
+        for (v = 0; v < ctx->ap_vg_len - 1; v++) {
```
Comment (tboinske-skyhook) [inline libel/beacons.c:169]: Why does the inner loop range over one fewer entries than the outer loop?

File: libel/beacons.c [hunk 12]
```diff
@@ -45,29 +45,143 @@ static bool beacon_in_cache(Sky_ctx_t *ctx, Beacon_t *b, Sky_cacheline_t *cl, in
...
+    }
+    if (!dup && v == ctx->ap_vg_len - 1) /* Not room for one more */
+        return false;
+
+    if (!dup) {
+        pvg[v] = delta;
+        ctx->beacon[vg].ap.vg_len = v + 1;
+    }
+
+    /* Add any virtual APs from the one being added (if it is a group) */
```
[the remark this answers is not in the document]
Comment (tboinske-skyhook) [reply libel/beacons.c:167]: Anyway, the intent here isn't obvious. Are you trying to handle the case where the child AP is itself the parent of a group? Can the comment describe what should be the final arrangement in this case?

File: libel/beacons.c (hunk 1 above)
Comment (gcleary) [reply libel/beacons.c:48]: Yup. I'll make it more clear.

**Comment by gcleary** [comment]
logging of the encoded request looks like this:

```
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() length 18
Skyhook libEL DEBG: utilities.c:get_vap_data() idx: 0
Skyhook libEL DEBG: utilities.c:get_vap_data() AP: 1 idx: 0 len: 2 ap: 1
Skyhook libEL DEBG: utilities.c:get_vap_data() 0000000: 01 61
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() string length 2
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() 0000000: 01 61
Skyhook libEL DEBG: utilities.c:get_vap_data() idx: 1
Skyhook libEL DEBG: utilities.c:get_vap_data() AP: 4 idx: 1 len: 2 ap: 4
Skyhook libEL DEBG: utilities.c:get_vap_data() 0000000: 04 D1
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() string length 2
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() 0000000: 04 D1
Skyhook libEL DEBG: utilities.c:get_vap_data() idx: 2
Skyhook libEL DEBG: utilities.c:get_vap_data() AP: 6 idx: 2 len: 2 ap: 6
Skyhook libEL DEBG: utilities.c:get_vap_data() 0000000: 06 C1
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() string length 2
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() 0000000: 06 C1
Skyhook libEL DEBG: utilities.c:get_vap_data() idx: 3
Skyhook libEL DEBG: utilities.c:get_vap_data() AP: 8 idx: 3 len: 5 ap: 8
Skyhook libEL DEBG: utilities.c:get_vap_data() 0000000: 08 3A 5A AA 6A
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() string length 5
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() 0000000: 08 3A 5A AA 6A
Skyhook libEL DEBG: utilities.c:get_vap_data() idx: 4
Skyhook libEL DEBG: utilities.c:get_vap_data() AP: 9 idx: 4 len: 2 ap: 9
Skyhook libEL DEBG: utilities.c:get_vap_data() 0000000: 09 4B
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() string length 2
Skyhook libEL DEBG: proto.c:encode_repeated_vap_() 0000000: 09 4B
Skyhook libEL DEBG: proto.c:serialize_request() 0000000: 0A 0C 12 34 56 12 34 56 11 22 33 44 55 66 10 C0
Skyhook libEL DEBG: proto.c:serialize_request() 0000010: A5 90 F6 05 1A 5B 08 01 22 45 DD 9C 82 97 CD 04
Skyhook libEL DEBG: proto.c:serialize_request() 0000020: B8 AA C5 E1 BA D9 38 99 B9 CB 84 AB CD 20 99 99
Skyhook libEL DEBG: proto.c:serialize_request() 0000030: CB 8C 9B ED 20 A0 A4 91 F5 AF 90 2F A2 C4 91 F5
Skyhook libEL DEBG: proto.c:serialize_request() 0000040: AF D0 2D B2 84 F3 90 B8 87 0A AA E2 F1 90 B8 87
Skyhook libEL DEBG: proto.c:serialize_request() 0000050: 0A 91 C4 91 F5 AF 90 2D A2 C4 91 F5 AF 90 2D 10
Skyhook libEL DEBG: proto.c:serialize_request() 0000060: CD 1C 32 0A 5C 01 01 5A 5B 52 4A 4D 4C 40 18 AD
Skyhook libEL DEBG: proto.c:serialize_request() 0000070: 02 22 02 0A 12 32 16 2A 02 E5 7D 32 02 F2 74 1A
Skyhook libEL DEBG: proto.c:serialize_request() 0000080: 02 DB 04 22 01 01 3A 01 64 42 02 BB 02 52 23 0A
Skyhook libEL DEBG: proto.c:serialize_request() 0000090: 04 BC B7 C2 11 12 04 88 91 BA 01 1A 01 6C 22 02
Skyhook libEL DEBG: proto.c:serialize_request() 00000A0: 8E 11 2A 01 28 32 01 64 3A 02 8E 02 42 01 05 4A
Skyhook libEL DEBG: proto.c:serialize_request() 00000B0: 01 64 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

I can't decode with protoc. When I visually look at the encoding of vap, the length of 18 makes sense (0x12), but I don't see the expected bytes of the strings e.g. 01 61 (length 2)

**Comment by gcleary** [comment]
Obsolete branch. Addressing Compressed Virtual APs in EC-63.

**Pull request closed without merging**